src/mjolnir/graphfilter.cc

```diff
@@ -219,7 +393,359 @@ void FilterTiles(GraphReader& reader,
+                       const graph_tile_ptr& tile,
+                       const DirectedEdge* directededge) {
+
+  // Get the tile at the end node.  Skip if node is in another tile.
+  // mode_change is not set for end nodes that are in diff tiles
+  if (directededge->endnode().tile_value() == tile->header()->graphid().tile_value()) {
+
+    // original edge data.
+    auto edgeinfo = tile->edgeinfo(directededge);
+    const NodeInfo* en_info = tile->node(directededge->endnode().id());
+
+    // do we need to aggregate?
+    if (en_info->mode_change()) {
+      std::unordered_set<std::string> isos;
+      bool isForward = directededge->forward();
+      auto id = directededge->endnode();
+      if (!isForward) {
+        std::reverse(shape.begin(), shape.end());
+      }
+      // walk in the correct direction.
+      uint64_t wayid = edgeinfo.wayid();
+      if (Aggregate(id, reader, shape, en, from_node, wayid, isos, isForward)) {
+        aggregated++; // count the current edge
+        // flip the shape back for storing in edgeinfo
+        if (!isForward) {
+          std::reverse(shape.begin(), shape.end());
+        }
+      }
+    }
+  }
+}
+
+// If we cross into another country we can't aggregate the edges
+// as the access can be different in each country.  Also, bollards
+// and or gates could exist at the node blocking access.
+//
+// Also, we need to handle islands that we created by tossing the
+// pedestrian edges.  For example://
+// https://www.openstreetmap.org/way/993706522
+// https://www.openstreetmap.org/way/975845893
+// As of 01/15/2024 there are only ~180 of these.
+
+void ValidateData(GraphReader& reader,
+                  std::list<PointLL>& shape,
+                  GraphId& en,
+                  std::unordered_set<GraphId>& processed_nodes,
+                  std::unordered_set<uint64_t>& no_agg_ways,
+                  const GraphId& from_node,
+                  const graph_tile_ptr& tile,
+                  const DirectedEdge* directededge) {
+
+  // Get the tile at the end node.  Skip if node is in another tile.
+  // mode_change is not set for end nodes that are in diff tiles
+  if (directededge->endnode().tile_value() == tile->header()->graphid().tile_value()) {
+
+    // original edge data.
+    auto edgeinfo = tile->edgeinfo(directededge);
+    const NodeInfo* en_info = tile->node(directededge->endnode().id());
+    const NodeInfo* sn_info = tile->node(from_node);
+
+    if (en_info->mode_change()) {
+
+      // If this edge has special attributes, then we can't aggregate
+      if (!CanAggregate(directededge)) {
+        processed_nodes.insert(directededge->endnode());
+        no_agg_ways.insert(edgeinfo.wayid());
+        return;
+      }
+
+      std::unordered_set<std::string> isos;
+      bool isForward = directededge->forward();
+      auto id = directededge->endnode();
+
+      isos.insert(tile->admin(sn_info->admin_index())->country_iso()); // start node
```

Comment:
is country_iso enough or we'd even need to consider state boundaries? aren't there different legislations across US states as well?

also would keeping a set of `admin_index()` not be enough instead of the full iso code?

Reply:
I think that we are fine here.  I followed the same logic used in the graph validator and shortcut builder.  They only look at the country ISO for the end nodes.